### ISSUE TITLE
[FEATURE REQUEST]: UPS-RS Support

### ISSUE THREAD

**Issue by ntenenz** [MGHComputationalPathology/dicomweb-client#37]
With the emergence of the [IHE AIW-I profile](https://www.ihe.net/uploadedFiles/Documents/Radiology/IHE_RAD_Suppl_AIW-I.pdf), the role of [UPS-RS](https://www.dicomstandard.org/dicomweb/workflow-ups-rs) is poised to grow dramatically. While [dcm4chee](https://github.com/dcm4che/dcm4chee-arc-light/tree/master/dcm4chee-arc-ups-rs) supports the [provider interface](https://dcm4chee-arc-cs.readthedocs.io/en/latest/overview.html), support for the user interface is relatively rare in the DICOM community. Given the widespread use of python for ML algorithms, a python-based UPS-RS client implementation would be beneficial for many.

**Comment by hackermd** [comment]
I agree, that would be valuable addition. We would happily accept a pull request. Personally, I have limited experience with UPS-RS so far and will need to investigate a bit. Do you already have concrete ideas regarding the interface (method names, parameters, return values, etc.)?

**Comment by ntenenz** [comment]
The summary/descriptions found [here](https://www.dicomstandard.org/dicomweb/workflow-ups-rs/) may serve as a good starting point:
* `create_work_item`
* `update_work_item`
* `retrive_work_item`
* `cancel_work_item`
* ...

Alternatively, while not grammatically correct, merging "work item" into "workitem" may be a bit more visually appealing and more closely maps onto the `store_instances`, `retrieve_instance`, etc. forms already present in the client.

**Comment by ntenenz** [comment]
The summary/descriptions found [here](https://www.dicomstandard.org/dicomweb/workflow-ups-rs/) may serve as a good starting point:
* `create_work_item`
* `update_work_item`
* `retrive_work_item`
* `cancel_work_item`
* ...

Alternatively, while not grammatically correct, merging "work item" into "workitem" may be a bit more visually appealing and more closely maps onto the `s`, `retrieve_instance`, etc. forms already present in the client.

**Comment by sjswerdloff** [comment]
Is this still of interest?  The DIMSE flavour of UPS is used extensively in IHE-RO TDW-II, and there is interest in migrating from DIMSE to DICOMWeb in the Radiation Oncology world.  I've started investigating OpenAPI 3.0 (Interface first) implementation of DICOM Web (starting with STOW-RS), but my next step once I've sorted out the details is to work on UPS-RS.

**Comment by CPBridge** [comment]
Sorry for the silence, yes this would be of interest. I also have little experience with UPS though

**Comment by sjswerdloff** [comment]
I'm downloading the dcm4chee docker so I can test against that.
I have been working on both sides: provider and user.
The OpenAPI approach was a bust, switched to just UPS-RS, first with Flask and now with Falcon (Flask debug causes problems with websocket servers) and manually coding + LLM assist, and making good progress).
I see the style of implementation here, so I can convert to that.  But will probably be a few weeks before I'm ready to make commits.  Is it preferable to commit one transaction at a time, or just "go for gold" and submit the whole thing at once (including tests... which will then require running a docker container of dcm4chee as part of the CI).
